src/application-insights/azext_applicationinsights/custom.py

```diff
@@ -31,3 +32,67 @@ def get_metric(cmd, client, application, metric, start_time=None, end_time=None,
 
 def get_metrics_metadata(cmd, client, application, resource_group_name=None):
     return client.metrics.get_metadata(get_id_from_azure_resource(cmd.cli_ctx, application, resource_group=resource_group_name))
+
+
+def create_or_update_component(client, application, resource_group_name, location, tags=None, kind="web", application_type='web'):
+    from .vendored_sdks.mgmt_applicationinsights.models import ApplicationInsightsComponent
+    component = ApplicationInsightsComponent(location, kind, application_type=application_type, tags=tags)
+    return client.create_or_update(resource_group_name, application, component)
+
+
+def update_component(client, application, resource_group_name, kind=None):
+    existing_component = client.get(resource_group_name, application)
+    if kind:
+        existing_component.kind = kind
+    return client.create_or_update(resource_group_name, application, existing_component)
+
+
+def update_component_tags(client, application, resource_group_name, tags):
+    return client.update_tags(resource_group_name, application, tags)
+
+
+def get_component(client, application, resource_group_name):
+    return client.get(resource_group_name, application)
+
+
+def show_components(client, application=None, resource_group_name=None):
+    if application:
+        if resource_group_name:
+            return client.get(resource_group_name, application)
+        raise CLIError("Application provided without resource group. Either specify app with resource group, or remove app.")
+    if resource_group_name:
+        return client.list_by_resource_group(resource_group_name)
+    return client.list()
+
+
+def delete_component(client, application, resource_group_name):
+    return client.delete(resource_group_name, application)
+
+
+def create_api_key(cmd, client, application, resource_group_name, api_key, read_properties=None, write_properties=None):
+    from .vendored_sdks.mgmt_applicationinsights.models import APIKeyRequest
+    if read_properties is None:
+        read_properties = ['ReadTelemetry', 'AuthenticateSDKControlChannel']
+    if write_properties is None:
+        write_properties = ['WriteAnnotations']
+    linked_read_properties, linked_write_properties = get_linked_properties(cmd.cli_ctx, application, resource_group_name, read_properties, write_properties)
+    api_key_request = APIKeyRequest(api_key, linked_read_properties, linked_write_properties)
+    return client.create(resource_group_name, application, api_key_request)
+
+
+def show_api_key(client, application, resource_group_name, api_key=None):
+    if api_key is None:
+        return client.list(resource_group_name, application)
+    result = list(filter(lambda result: result.name == api_key, client.list(resource_group_name, application)))
+    if len(result) == 1:
+        return result[0]
```

Comment:
I suggest just return `result` to have consistent return type in array which would help the scripting